src/mango/src/mango_indexer.erl

```diff
@@ -0,0 +1,130 @@
+-include_lib("couch_views/include/couch_views.hrl").
+
+
+modify(Db, Change, Doc, PrevDoc, Seq) ->
+    try
+        Indexes = json_indexes(Db, Doc),
+        modify_int(Db, Change, Doc, Indexes, Seq)
+    catch
+        Error:Reason ->
+            DbName = fabric2_db:name(Db),
+            io:format("ERROR ~p ~p ~p ~n", [Error, Reason, erlang:display(erlang:get_stacktrace())]),
+            Id = doc_id(Doc, PrevDoc),
+            couch_log:error("Mango index error for Db ~s Doc ~p ~p ~p",
+                [DbName, Id, Error, Reason])
+    end.
+
+
+doc_id(undefined, #doc{id = DocId}) ->
+    DocId;
+doc_id(undefined, _) ->
+    <<"unknown_doc_id">>;
+doc_id(#doc{id = DocId}, _) ->
+    DocId.
+
+
+% Check if design doc is mango index and kick off background worker
+% to build the new index
+modify_int(Db, Change, #doc{id = <<?DESIGN_DOC_PREFIX, _/binary>>} = Doc,
+        Indexes, Seq) when Change == created orelse Change == updated ->
+    DbName = fabric2_db:name(Db),
+
+    {Props} = couch_doc:to_json_obj(Doc, []),
+    case proplists:get_value(<<"language">>, Props) of
+        <<"query">> ->
+            {ok, Mrst} = couch_mrview_util:ddoc_to_mrst(DbName, Doc),
+            couch_views_fdb:create_build_vs(Db, Mrst),
+            {ok, _} = couch_views_jobs:build_view_async(Db, Mrst, true);
+        _ ->
+            ok
+    end,
+    update_indexes_seq(Db, Indexes, Seq);
+
+modify_int(Db, _Change, Doc, Indexes, Seq) ->
+    write_doc(Db, Doc, Indexes, Seq).
+
+
+write_doc(Db, #doc{deleted = Deleted} = Doc, Indexes, Seq) ->
+    #doc{id = DocId} = Doc,
+    JsonDoc = mango_json:to_binary(couch_doc:to_json_obj(Doc, [])),
+
+    lists:foreach(fun (Idx) ->
+        DocResult0 = #{
+            id => DocId,
+            results => []
+        },
+
+        DocResult1 = case Deleted of
+            true ->
+                DocResult0#{deleted => true};
+            false ->
+                Results = mango_eval:index_doc([Idx], JsonDoc),
+                DocResult0#{results => Results}
+        end,
+
+        DbName = mango_idx:dbname(Idx),
+        DDoc = mango_idx:ddoc(Idx),
+        {ok, Mrst} = couch_mrview_util:ddoc_to_mrst(DbName, DDoc),
+        couch_views_indexer:write_docs(Db, Mrst, [DocResult1], #{last_seq => Seq})
+    end, Indexes).
+
+json_indexes(Db,  #doc{id = <<?DESIGN_DOC_PREFIX, _/binary>> = Id}) ->
+    json_indexes(Db, [{ignore, [Id]}]);
+
+json_indexes(Db,  #doc{}) ->
+    json_indexes(Db, []);
+
+json_indexes(Db, Opts) when is_list(Opts) ->
+    #{
+        md_version := MdVersion
+    } = Db,
+
+    case get(json_indexes) of
+        {SavedVersion, Indexes} when MdVersion == SavedVersion ->
+            Indexes;
+        _ ->
+            Indexes = lists:filter(fun(Idx) ->
+                Idx#idx.type == <<"json">>
+            end, mango_idx:list(Db, Opts)),
+            put(json_indexes, {MdVersion, Indexes}),
+            Indexes
+    end.
+
+
+update_indexes_seq(Db, Indexes, Seq) ->
```

Comment:
Am I reading this right? Any index that's currently build is extended to the current update_seq when any design document is updated? That logic seems overly specific and different from what I remember thinking the intent was (build to when a ddoc defines a new index and stop). Also seems odd that its all indexes and only on design doc updates?

Reply:
If we have a situation where we have say [doc1, doc2, ddoc] and they all saved to the database in different transactions. If we only updated the indexes for the normal docs and not the design docs then all the mango indexes would be behind by one document. Which would mean that a mango query happens we would need to build the index before reading the index. This way all indexes are always progressed when a design doc is saved.

Reply:
This section caused the most hassle to get correct. I would be happy to leave it out and rather that we do build the index slightly if its behind based on a few design docs.

Reply:
So, to make sure I have that, the issue is that when we save a design doc, that design doc is not processed by other indexes. And given that our views aren't considered up to date until all changes are processed, and those views don't update past their update_seq, it means that we get stuck and can't answer queries against those view indexes?

And so to avoid that, this just bumps the update_seq for all views that are already built? That seems awfully sketchy. For instance, it'll leave any index that's currently building broken until the first design doc update after it finishes building. Second, it seems a bit strange to be checkpointing that update_seq from oustdie the couch_views_indexer process which has a lot of work in guaranteeing its the only thing touching that value at any given time.

Seems more like we should be looking at rearranging the logic a bit such that we're able to make that update without having to worry about corrupting the index somehow.

Reply:
No that is not quite correct. For a database if the last doc that is added to the database is a design doc and we don't call this function, then all indexes are behind by one document. So when a mango query is run, it will have to kick off a background job to process that design doc and set the update_seq to the final change before querying the document. 

There is some subtlety around managing the indexes update_seq. The logic works as follows:
* while couch_views_indexer is building the index, it is the only process updating the update_seq
* mango_indexer will continue to index any new documents but will not update the update_seq
* Once the couch_views_indexer completes building to the build_vs, it will do a final update_seq to the latest seq in the database. This is because we assuming that if you using build_vs that the index is also being updated in a doc transaction
* Only once the index is built to the build_vs then the mango_indexer will start updating the update_seq so that the index update_seq is up to date with the indexing.